Add Semantic Version field type mapper and extensive unit tests

deshsidd opened the pull request:
Add Semantic Version field type mapper and extensive unit tests

This PR introduces a new version field type that follows [Semantic Versioning (SemVer)](https://semver.org/) semantics. It allows indexing and querying of version strings such as 1.0.0, 1.0.0-alpha, 2.1.0+build.1, etc., with correct ordering and range support.

## Description
Summary:
- Implements a SemanticVersionFieldMapper that extends TermBasedFieldType.
- Internally normalizes semantic version strings into sortable format.
- Supports common query types like exists, match, multi-match, term, terms, range, prefix, wildcard, regex, fuzzy, ids query, constant score, function score, match phrase, sort, etc.
- Build metadata (e.g., +build123) is ignored for ordering per SemVer spec.
- Pre-releases (e.g., -alpha, -rc.1) are handled correctly during comparisons.

Code Additions:
- SemanticVersion.java: Semantic version model and comparator logic
- SemanticVersionFieldMapper.java: Mapper definition and version field parser
- Registered in IndicesModule.java

Tests:
Includes full unit test coverage. 
- Parsing and normalization: SemanticVersionTests.java
- Mapper behavior and indexing: SemanticVersionFieldMapperTests.java
- Query support across various query types: SemanticVersionFieldQueryTests.java

## Related Issues
Addresses #16814

## Check List
- [ ] Functionality includes testing.
- [ ] API changes companion pull request [created](https://github.com/opensearch-project/opensearch-api-specification/blob/main/DEVELOPER_GUIDE.md), if applicable.
- [ ] Public documentation issue/PR [created](https://github.com/opensearch-project/documentation-website/issues/new/choose), if applicable.

## Examples
**1. Create index with mapping**

Request:
```
PUT test_versions
{
  "mappings": {
    "properties": {
      "app": {
        "type": "keyword"
      },
      "version": {
        "type": "version"
      },
      "release_date": {
        "type": "date"
      },
      "description": {
        "type": "text"
      }
    }
  }
}
```
Response:
```
{
  "acknowledged": true,
  "shards_acknowledged": true,
  "index": "test_versions"
}
```

**2. Index data**

Request:
```
POST test_versions/_bulk
{ "index": {} }
{ "app": "AlphaApp", "version": "1.0.0", "release_date": "2023-01-01", "description": "Initial release" }
{ "index": {} }
{ "app": "AlphaApp", "version": "1.0.1", "release_date": "2023-02-15", "description": "Bug fix release" }
{ "index": {} }
{ "app": "AlphaApp", "version": "1.1.0", "release_date": "2023-05-10", "description": "Minor feature update" }
{ "index": {} }
{ "app": "AlphaApp", "version": "2.0.0", "release_date": "2024-01-01", "description": "Major release" }
{ "index": {} }
{ "app": "BetaApp",  "version": "0.9.0", "release_date": "2022-12-01", "description": "Beta release" }
{ "index": {} }
{ "app": "BetaApp",  "version": "1.0.0-alpha", "release_date": "2023-03-01", "description": "Alpha pre-release" }
{ "index": {} }
{ "app": "BetaApp",  "version": "1.0.0-alpha.1", "release_date": "2023-03-10", "description": "Alpha patch" }
{ "index": {} }
{ "app": "BetaApp",  "version": "1.0.0-beta", "release_date": "2023-04-01", "description": "Beta pre-release" }
{ "index": {} }
{ "app": "BetaApp",  "version": "1.0.0-rc.1", "release_date": "2023-04-15", "description": "Release candidate" }
{ "index": {} }
{ "app": "BetaApp",  "version": "1.0.0+20230815", "release_date": "2023-08-15", "description": "Build metadata release" }
{ "index": {} }
{ "app": "GammaApp", "version": "1.2.0", "release_date": "2023-09-01", "description": "Stable release" }
{ "index": {} }
{ "app": "GammaApp", "version": "1.2.0-alpha", "release_date": "2023-07-15", "description": "Alpha version of 1.2" }
{ "index": {} }
{ "app": "GammaApp", "version": "1.2.0-beta.2", "release_date": "2023-08-01", "description": "Beta 2 version of 1.2" }
{ "index": {} }
{ "app": "GammaApp", "version": "1.2.0-rc.1", "release_date": "2023-08-15", "description": "RC version of 1.2" }
```
Response:
```
{
  "took": 23,
  "errors": false,
  "items": [
    {
      "index": {
        "_index": "test_versions",
        "_id": "cCWNQpcBY7cEASBv3Vr1",
        "_version": 1,
        "result": "created",
        "_shards": {
          "total": 2,
          "successful": 1,
          "failed": 0
        },
        "_seq_no": 0,
        "_primary_term": 1,
        "status": 201
      }
    },
    {
      "index": {
        "_index": "test_versions",
        "_id": "cSWNQpcBY7cEASBv3Vr1",
        "_version": 1,
        "result": "created",
        "_shards": {
          "total": 2,
          "successful": 1,
          "failed": 0
        },
        "_seq_no": 1,
        "_primary_term": 1,
        "status": 201
      }
    },
    {
      "index": {
        "_index": "test_versions",
        "_id": "ciWNQpcBY7cEASBv3Vr1",
        "_version": 1,
        "result": "created",
        "_shards": {
          "total": 2,
          "successful": 1,
          "failed": 0
        },
        "_seq_no": 2,
        "_primary_term": 1,
        "status": 201
      }
    },
    {
      "index": {
        "_index": "test_versions",
        "_id": "cyWNQpcBY7cEASBv3Vr1",
        "_version": 1,
        "result": "created",
        "_shards": {
          "total": 2,
          "successful": 1,
          "failed": 0
        },
        "_seq_no": 3,
        "_primary_term": 1,
        "status": 201
      }
    },
    {
      "index": {
        "_index": "test_versions",
        "_id": "dCWNQpcBY7cEASBv3Vr1",
        "_version": 1,
        "result": "created",
        "_shards": {
          "total": 2,
          "successful": 1,
          "failed": 0
        },
        "_seq_no": 4,
        "_primary_term": 1,
        "status": 201
      }
    },
    {
      "index": {
        "_index": "test_versions",
        "_id": "dSWNQpcBY7cEASBv3Vr1",
        "_version": 1,
        "result": "created",
        "_shards": {
          "total": 2,
          "successful": 1,
          "failed": 0
        },
        "_seq_no": 5,
        "_primary_term": 1,
        "status": 201
      }
    },
    {
      "index": {
        "_index": "test_versions",
        "_id": "diWNQpcBY7cEASBv3Vr1",
        "_version": 1,
        "result": "created",
        "_shards": {
          "total": 2,
          "successful": 1,
          "failed": 0
        },
        "_seq_no": 6,
        "_primary_term": 1,
        "status": 201
      }
    },
    {
      "index": {
        "_index": "test_versions",
        "_id": "dyWNQpcBY7cEASBv3Vr1",
        "_version": 1,
        "result": "created",
        "_shards": {
          "total": 2,
          "successful": 1,
          "failed": 0
        },
        "_seq_no": 7,
        "_primary_term": 1,
        "status": 201
      }
    },
    {
      "index": {
        "_index": "test_versions",
        "_id": "eCWNQpcBY7cEASBv3Vr1",
        "_version": 1,
        "result": "created",
        "_shards": {
          "total": 2,
          "successful": 1,
          "failed": 0
        },
        "_seq_no": 8,
        "_primary_term": 1,
        "status": 201
      }
    },
    {
      "index": {
        "_index": "test_versions",
        "_id": "eSWNQpcBY7cEASBv3Vr1",
        "_version": 1,
        "result": "created",
        "_shards": {
          "total": 2,
          "successful": 1,
          "failed": 0
        },
        "_seq_no": 9,
        "_primary_term": 1,
        "status": 201
      }
    },
    {
      "index": {
        "_index": "test_versions",
        "_id": "eiWNQpcBY7cEASBv3Vr1",
        "_version": 1,
        "result": "created",
        "_shards": {
          "total": 2,
          "successful": 1,
          "failed": 0
        },
        "_seq_no": 10,
        "_primary_term": 1,
        "status": 201
      }
    },
    {
      "index": {
        "_index": "test_versions",
        "_id": "eyWNQpcBY7cEASBv3Vr1",
        "_version": 1,
        "result": "created",
        "_shards": {
          "total": 2,
          "successful": 1,
          "failed": 0
        },
        "_seq_no": 11,
        "_primary_term": 1,
        "status": 201
      }
    },
    {
      "index": {
        "_index": "test_versions",
        "_id": "fCWNQpcBY7cEASBv3Vr1",
        "_version": 1,
        "result": "created",
        "_shards": {
          "total": 2,
          "successful": 1,
          "failed": 0
        },
        "_seq_no": 12,
        "_primary_term": 1,
        "status": 201
      }
    },
    {
      "index": {
        "_index": "test_versions",
        "_id": "fSWNQpcBY7cEASBv3Vr1",
        "_version": 1,
        "result": "created",
        "_shards": {
          "total": 2,
          "successful": 1,
          "failed": 0
        },
        "_seq_no": 13,
        "_primary_term": 1,
        "status": 201
      }
    }
  ]
}
```

**3a. match_all query**

Request:
```
GET test_versions/_search
{
  "query": { "match_all": {} }
}
```
Response:
```
{
  "took": 1014,
  "timed_out": false,
  "_shards": {
    "total": 1,
    "successful": 1,
    "skipped": 0,
    "failed": 0
  },
  "hits": {
    "total": {
      "value": 14,
      "relation": "eq"
    },
    "max_score": 1,
    "hits": [
      {
        "_index": "test_versions",
        "_id": "cCWNQpcBY7cEASBv3Vr1",
        "_score": 1,
        "_source": {
          "app": "AlphaApp",
          "version": "1.0.0",
          "release_date": "2023-01-01",
          "description": "Initial release"
        }
      },
      {
        "_index": "test_versions",
        "_id": "cSWNQpcBY7cEASBv3Vr1",
        "_score": 1,
        "_source": {
          "app": "AlphaApp",
          "version": "1.0.1",
          "release_date": "2023-02-15",
          "description": "Bug fix release"
        }
      },
      {
        "_index": "test_versions",
        "_id": "ciWNQpcBY7cEASBv3Vr1",
        "_score": 1,
        "_source": {
          "app": "AlphaApp",
          "version": "1.1.0",
          "release_date": "2023-05-10",
          "description": "Minor feature update"
        }
      },
      {
        "_index": "test_versions",
        "_id": "cyWNQpcBY7cEASBv3Vr1",
        "_score": 1,
        "_source": {
          "app": "AlphaApp",
          "version": "2.0.0",
          "release_date": "2024-01-01",
          "description": "Major release"
        }
      },
      {
        "_index": "test_versions",
        "_id": "dCWNQpcBY7cEASBv3Vr1",
        "_score": 1,
        "_source": {
          "app": "BetaApp",
          "version": "0.9.0",
          "release_date": "2022-12-01",
          "description": "Beta release"
        }
      },
      {
        "_index": "test_versions",
        "_id": "dSWNQpcBY7cEASBv3Vr1",
        "_score": 1,
        "_source": {
          "app": "BetaApp",
          "version": "1.0.0-alpha",
          "release_date": "2023-03-01",
          "description": "Alpha pre-release"
        }
      },
      {
        "_index": "test_versions",
        "_id": "diWNQpcBY7cEASBv3Vr1",
        "_score": 1,
        "_source": {
          "app": "BetaApp",
          "version": "1.0.0-alpha.1",
          "release_date": "2023-03-10",
          "description": "Alpha patch"
        }
      },
      {
        "_index": "test_versions",
        "_id": "dyWNQpcBY7cEASBv3Vr1",
        "_score": 1,
        "_source": {
          "app": "BetaApp",
          "version": "1.0.0-beta",
          "release_date": "2023-04-01",
          "description": "Beta pre-release"
        }
      },
      {
        "_index": "test_versions",
        "_id": "eCWNQpcBY7cEASBv3Vr1",
        "_score": 1,
        "_source": {
          "app": "BetaApp",
          "version": "1.0.0-rc.1",
          "release_date": "2023-04-15",
          "description": "Release candidate"
        }
      },
      {
        "_index": "test_versions",
        "_id": "eSWNQpcBY7cEASBv3Vr1",
        "_score": 1,
        "_source": {
          "app": "BetaApp",
          "version": "1.0.0+20230815",
          "release_date": "2023-08-15",
          "description": "Build metadata release"
        }
      }
    ]
  }
}
```

**3b. term query**

Request:
```
GET test_versions/_search
{
  "query": {
    "term": { "version": "1.0.1" }
  }
}
```
Response:
```
{
  "took": 5,
  "timed_out": false,
  "_shards": {
    "total": 1,
    "successful": 1,
    "skipped": 0,
    "failed": 0
  },
  "hits": {
    "total": {
      "value": 1,
      "relation": "eq"
    },
    "max_score": 1.0466295,
    "hits": [
      {
        "_index": "test_versions",
        "_id": "cSWNQpcBY7cEASBv3Vr1",
        "_score": 1.0466295,
        "_source": {
          "app": "AlphaApp",
          "version": "1.0.1",
          "release_date": "2023-02-15",
          "description": "Bug fix release"
        }
      }
    ]
  }
}
```

**3c. terms query**

Request:
```
GET test_versions/_search
{
  "query": {
    "terms": {
      "version": ["1.0.0", "1.2.0-alpha"]
    }
  }
}

```

Response:
```
{
  "took": 26,
  "timed_out": false,
  "_shards": {
    "total": 1,
    "successful": 1,
    "skipped": 0,
    "failed": 0
  },
  "hits": {
    "total": {
      "value": 2,
      "relation": "eq"
    },
    "max_score": 1,
    "hits": [
      {
        "_index": "test_versions",
        "_id": "cCWNQpcBY7cEASBv3Vr1",
        "_score": 1,
        "_source": {
          "app": "AlphaApp",
          "version": "1.0.0",
          "release_date": "2023-01-01",
          "description": "Initial release"
        }
      },
      {
        "_index": "test_versions",
        "_id": "eyWNQpcBY7cEASBv3Vr1",
        "_score": 1,
        "_source": {
          "app": "GammaApp",
          "version": "1.2.0-alpha",
          "release_date": "2023-07-15",
          "description": "Alpha version of 1.2"
        }
      }
    ]
  }
}
``` 


**3d. range query**

Request:
```
GET test_versions/_search
{
  "query": {
    "range": {
      "version": { "gte": "2.0.0" }
    }
  }
}
```

Response:
```
{
  "took": 3,
  "timed_out": false,
  "_shards": {
    "total": 1,
    "successful": 1,
    "skipped": 0,
    "failed": 0
  },
  "hits": {
    "total": {
      "value": 1,
      "relation": "eq"
    },
    "max_score": 1,
    "hits": [
      {
        "_index": "test_versions",
        "_id": "cyWNQpcBY7cEASBv3Vr1",
        "_score": 1,
        "_source": {
          "app": "AlphaApp",
          "version": "2.0.0",
          "release_date": "2024-01-01",
          "description": "Major release"
        }
      }
    ]
  }
}
```

**3e. exists query**

Request:
```
GET test_versions/_search
{
  "query": {
    "exists": {
      "field": "version"
    }
  }
}
```

Response:
```
{
  "took": 2,
  "timed_out": false,
  "_shards": {
    "total": 1,
    "successful": 1,
    "skipped": 0,
    "failed": 0
  },
  "hits": {
    "total": {
      "value": 14,
      "relation": "eq"
    },
    "max_score": 1,
    "hits": [
      {
        "_index": "test_versions",
        "_id": "cCWNQpcBY7cEASBv3Vr1",
        "_score": 1,
        "_source": {
          "app": "AlphaApp",
          "version": "1.0.0",
          "release_date": "2023-01-01",
          "description": "Initial release"
        }
      },
      {
        "_index": "test_versions",
        "_id": "cSWNQpcBY7cEASBv3Vr1",
        "_score": 1,
        "_source": {
          "app": "AlphaApp",
          "version": "1.0.1",
          "release_date": "2023-02-15",
          "description": "Bug fix release"
        }
      },
      {
        "_index": "test_versions",
        "_id": "ciWNQpcBY7cEASBv3Vr1",
        "_score": 1,
        "_source": {
          "app": "AlphaApp",
          "version": "1.1.0",
          "release_date": "2023-05-10",
          "description": "Minor feature update"
        }
      },
      {
        "_index": "test_versions",
        "_id": "cyWNQpcBY7cEASBv3Vr1",
        "_score": 1,
        "_source": {
          "app": "AlphaApp",
          "version": "2.0.0",
          "release_date": "2024-01-01",
          "description": "Major release"
        }
      },
      {
        "_index": "test_versions",
        "_id": "dCWNQpcBY7cEASBv3Vr1",
        "_score": 1,
        "_source": {
          "app": "BetaApp",
          "version": "0.9.0",
          "release_date": "2022-12-01",
          "description": "Beta release"
        }
      },
      {
        "_index": "test_versions",
        "_id": "dSWNQpcBY7cEASBv3Vr1",
        "_score": 1,
        "_source": {
          "app": "BetaApp",
          "version": "1.0.0-alpha",
          "release_date": "2023-03-01",
          "description": "Alpha pre-release"
        }
      },
      {
        "_index": "test_versions",
        "_id": "diWNQpcBY7cEASBv3Vr1",
        "_score": 1,
        "_source": {
          "app": "BetaApp",
          "version": "1.0.0-alpha.1",
          "release_date": "2023-03-10",
          "description": "Alpha patch"
        }
      },
      {
        "_index": "test_versions",
        "_id": "dyWNQpcBY7cEASBv3Vr1",
        "_score": 1,
        "_source": {
          "app": "BetaApp",
          "version": "1.0.0-beta",
          "release_date": "2023-04-01",
          "description": "Beta pre-release"
        }
      },
      {
        "_index": "test_versions",
        "_id": "eCWNQpcBY7cEASBv3Vr1",
        "_score": 1,
        "_source": {
          "app": "BetaApp",
          "version": "1.0.0-rc.1",
          "release_date": "2023-04-15",
          "description": "Release candidate"
        }
      },
      {
        "_index": "test_versions",
        "_id": "eSWNQpcBY7cEASBv3Vr1",
        "_score": 1,
        "_source": {
          "app": "BetaApp",
          "version": "1.0.0+20230815",
          "release_date": "2023-08-15",
          "description": "Build metadata release"
        }
      }
    ]
  }
}
```

**3f. match query**

Request:
```
GET test_versions/_search
{
  "query": {
    "match": {
      "version": "1.0.0"
    }
  }
}
```

Response:
```
{
  "took": 16,
  "timed_out": false,
  "_shards": {
    "total": 1,
    "successful": 1,
    "skipped": 0,
    "failed": 0
  },
  "hits": {
    "total": {
      "value": 1,
      "relation": "eq"
    },
    "max_score": 1.0466295,
    "hits": [
      {
        "_index": "test_versions",
        "_id": "cCWNQpcBY7cEASBv3Vr1",
        "_score": 1.0466295,
        "_source": {
          "app": "AlphaApp",
          "version": "1.0.0",
          "release_date": "2023-01-01",
          "description": "Initial release"
        }
      }
    ]
  }
}
```

**3g. multi_match query**

Request:
```
GET test_versions/_search
{
  "query": {
    "multi_match": {
      "query": "1.2.0-alpha.1",
      "fields": ["version", "description"]
    }
  }
}
```

Response:
```
{
  "took": 6,
  "timed_out": false,
  "_shards": {
    "total": 1,
    "successful": 1,
    "skipped": 0,
    "failed": 0
  },
  "hits": {
    "total": {
      "value": 3,
      "relation": "eq"
    },
    "max_score": 0.7540348,
    "hits": [
      {
        "_index": "test_versions",
        "_id": "diWNQpcBY7cEASBv3Vr1",
        "_score": 0.7540348,
        "_source": {
          "app": "BetaApp",
          "version": "1.0.0-alpha.1",
          "release_date": "2023-03-10",
          "description": "Alpha patch"
        }
      },
      {
        "_index": "test_versions",
        "_id": "dSWNQpcBY7cEASBv3Vr1",
        "_score": 0.64823484,
        "_source": {
          "app": "BetaApp",
          "version": "1.0.0-alpha",
          "release_date": "2023-03-01",
          "description": "Alpha pre-release"
        }
      },
      {
        "_index": "test_versions",
        "_id": "eyWNQpcBY7cEASBv3Vr1",
        "_score": 0.5684715,
        "_source": {
          "app": "GammaApp",
          "version": "1.2.0-alpha",
          "release_date": "2023-07-15",
          "description": "Alpha version of 1.2"
        }
      }
    ]
  }
}
```

**3h. prefix query**

Request:
```
GET test_versions/_search
{
  "query": {
    "prefix": {
      "version": {
        "value": "1.2.0-a"
      }
    }
  }
}
```

Response:
```
{
  "took": 2,
  "timed_out": false,
  "_shards": {
    "total": 1,
    "successful": 1,
    "skipped": 0,
    "failed": 0
  },
  "hits": {
    "total": {
      "value": 1,
      "relation": "eq"
    },
    "max_score": 1,
    "hits": [
      {
        "_index": "test_versions",
        "_id": "eyWNQpcBY7cEASBv3Vr1",
        "_score": 1,
        "_source": {
          "app": "GammaApp",
          "version": "1.2.0-alpha",
          "release_date": "2023-07-15",
          "description": "Alpha version of 1.2"
        }
      }
    ]
  }
}
```

**3i. wildcard query**

Request:
```
GET test_versions/_search
{
  "query": {
    "wildcard": {
      "version": "1.2.0-*"
    }
  }
}
```

Response:
```
{
  "took": 5,
  "timed_out": false,
  "_shards": {
    "total": 1,
    "successful": 1,
    "skipped": 0,
    "failed": 0
  },
  "hits": {
    "total": {
      "value": 3,
      "relation": "eq"
    },
    "max_score": 1,
    "hits": [
      {
        "_index": "test_versions",
        "_id": "eyWNQpcBY7cEASBv3Vr1",
        "_score": 1,
        "_source": {
          "app": "GammaApp",
          "version": "1.2.0-alpha",
          "release_date": "2023-07-15",
          "description": "Alpha version of 1.2"
        }
      },
      {
        "_index": "test_versions",
        "_id": "fCWNQpcBY7cEASBv3Vr1",
        "_score": 1,
        "_source": {
          "app": "GammaApp",
          "version": "1.2.0-beta.2",
          "release_date": "2023-08-01",
          "description": "Beta 2 version of 1.2"
        }
      },
      {
        "_index": "test_versions",
        "_id": "fSWNQpcBY7cEASBv3Vr1",
        "_score": 1,
        "_source": {
          "app": "GammaApp",
          "version": "1.2.0-rc.1",
          "release_date": "2023-08-15",
          "description": "RC version of 1.2"
        }
      }
    ]
  }
}
```

**3j. regex query**

Request:
```
GET test_versions/_search
{
  "query": {
    "regexp": {
      "version": "1\\.2\\.0\\-[a-z]+\\.[0-9]+"
    }
  }
}
```

Response:
```
{
  "took": 8,
  "timed_out": false,
  "_shards": {
    "total": 1,
    "successful": 1,
    "skipped": 0,
    "failed": 0
  },
  "hits": {
    "total": {
      "value": 2,
      "relation": "eq"
    },
    "max_score": 1,
    "hits": [
      {
        "_index": "test_versions",
        "_id": "fCWNQpcBY7cEASBv3Vr1",
        "_score": 1,
        "_source": {
          "app": "GammaApp",
          "version": "1.2.0-beta.2",
          "release_date": "2023-08-01",
          "description": "Beta 2 version of 1.2"
        }
      },
      {
        "_index": "test_versions",
        "_id": "fSWNQpcBY7cEASBv3Vr1",
        "_score": 1,
        "_source": {
          "app": "GammaApp",
          "version": "1.2.0-rc.1",
          "release_date": "2023-08-15",
          "description": "RC version of 1.2"
        }
      }
    ]
  }
}
```

**3k. match_phrase query**

Request:
```
GET test_versions/_search
{
  "query": {
    "match_phrase": {
      "version": "1.2.0-beta.2"
    }
  }
}
```

Response:
```
{
  "took": 2,
  "timed_out": false,
  "_shards": {
    "total": 1,
    "successful": 1,
    "skipped": 0,
    "failed": 0
  },
  "hits": {
    "total": {
      "value": 1,
      "relation": "eq"
    },
    "max_score": 1.0466295,
    "hits": [
      {
        "_index": "test_versions",
        "_id": "fCWNQpcBY7cEASBv3Vr1",
        "_score": 1.0466295,
        "_source": {
          "app": "GammaApp",
          "version": "1.2.0-beta.2",
          "release_date": "2023-08-01",
          "description": "Beta 2 version of 1.2"
        }
      }
    ]
  }
}
```

**3l. sort query**

Request:
```
GET test_versions/_search
{
  "query": { "match_all": {} },
  "sort": [
    { "version": { "order": "asc" }}
  ]
}
```

Response:
```
{
  "took": 3,
  "timed_out": false,
  "_shards": {
    "total": 1,
    "successful": 1,
    "skipped": 0,
    "failed": 0
  },
  "hits": {
    "total": {
      "value": 14,
      "relation": "eq"
    },
    "max_score": null,
    "hits": [
      {
        "_index": "test_versions",
        "_id": "dCWNQpcBY7cEASBv3Vr1",
        "_score": null,
        "_source": {
          "app": "BetaApp",
          "version": "0.9.0",
          "release_date": "2022-12-01",
          "description": "Beta release"
        },
        "sort": [
          "00000000000000000000.00000000000000000009.00000000000000000000~"
        ]
      },
      {
        "_index": "test_versions",
        "_id": "dSWNQpcBY7cEASBv3Vr1",
        "_score": null,
        "_source": {
          "app": "BetaApp",
          "version": "1.0.0-alpha",
          "release_date": "2023-03-01",
          "description": "Alpha pre-release"
        },
        "sort": [
          "00000000000000000001.00000000000000000000.00000000000000000000-alpha"
        ]
      },
      {
        "_index": "test_versions",
        "_id": "diWNQpcBY7cEASBv3Vr1",
        "_score": null,
        "_source": {
          "app": "BetaApp",
          "version": "1.0.0-alpha.1",
          "release_date": "2023-03-10",
          "description": "Alpha patch"
        },
        "sort": [
          "00000000000000000001.00000000000000000000.00000000000000000000-alpha.1"
        ]
      },
      {
        "_index": "test_versions",
        "_id": "dyWNQpcBY7cEASBv3Vr1",
        "_score": null,
        "_source": {
          "app": "BetaApp",
          "version": "1.0.0-beta",
          "release_date": "2023-04-01",
          "description": "Beta pre-release"
        },
        "sort": [
          "00000000000000000001.00000000000000000000.00000000000000000000-beta"
        ]
      },
      {
        "_index": "test_versions",
        "_id": "eCWNQpcBY7cEASBv3Vr1",
        "_score": null,
        "_source": {
          "app": "BetaApp",
          "version": "1.0.0-rc.1",
          "release_date": "2023-04-15",
          "description": "Release candidate"
        },
        "sort": [
          "00000000000000000001.00000000000000000000.00000000000000000000-rc.1"
        ]
      },
      {
        "_index": "test_versions",
        "_id": "cCWNQpcBY7cEASBv3Vr1",
        "_score": null,
        "_source": {
          "app": "AlphaApp",
          "version": "1.0.0",
          "release_date": "2023-01-01",
          "description": "Initial release"
        },
        "sort": [
          "00000000000000000001.00000000000000000000.00000000000000000000~"
        ]
      },
      {
        "_index": "test_versions",
        "_id": "eSWNQpcBY7cEASBv3Vr1",
        "_score": null,
        "_source": {
          "app": "BetaApp",
          "version": "1.0.0+20230815",
          "release_date": "2023-08-15",
          "description": "Build metadata release"
        },
        "sort": [
          "00000000000000000001.00000000000000000000.00000000000000000000~"
        ]
      },
      {
        "_index": "test_versions",
        "_id": "cSWNQpcBY7cEASBv3Vr1",
        "_score": null,
        "_source": {
          "app": "AlphaApp",
          "version": "1.0.1",
          "release_date": "2023-02-15",
          "description": "Bug fix release"
        },
        "sort": [
          "00000000000000000001.00000000000000000000.00000000000000000001~"
        ]
      },
      {
        "_index": "test_versions",
        "_id": "ciWNQpcBY7cEASBv3Vr1",
        "_score": null,
        "_source": {
          "app": "AlphaApp",
          "version": "1.1.0",
          "release_date": "2023-05-10",
          "description": "Minor feature update"
        },
        "sort": [
          "00000000000000000001.00000000000000000001.00000000000000000000~"
        ]
      },
      {
        "_index": "test_versions",
        "_id": "eyWNQpcBY7cEASBv3Vr1",
        "_score": null,
        "_source": {
          "app": "GammaApp",
          "version": "1.2.0-alpha",
          "release_date": "2023-07-15",
          "description": "Alpha version of 1.2"
        },
        "sort": [
          "00000000000000000001.00000000000000000002.00000000000000000000-alpha"
        ]
      }
    ]
  }
}
```



By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

deshsidd commented:
Please re-trigger the build and help with reviewing this PR.

cc @cwperks @jed326 @mch2 @harshavamsi @rishabhmaurya @soosinha @rafael-gumiero

cwperks reviewed (approved):
This change LGTM! Can another maintainer please review this one too?

harshavamsi commented on server/src/main/java/org/opensearch/index/mapper/SemanticVersionFieldMapper.java:
```diff
@@ -0,0 +1,313 @@
+        @Override
+        public Query wildcardQuery(String value, MultiTermQuery.RewriteMethod method, boolean caseInsensitive, QueryShardContext context) {
+            if (caseInsensitive) {
+                value = value.toLowerCase(Locale.ROOT);
+            }
+            return new WildcardQuery(new Term(name(), indexedValueForSearch(value)), Operations.DEFAULT_DETERMINIZE_WORK_LIMIT);
+        }
+
+        @Override
+        public Query prefixQuery(String value, MultiTermQuery.RewriteMethod method, boolean caseInsensitive, QueryShardContext context) {
+            if (method == null) {
+                method = MultiTermQuery.CONSTANT_SCORE_REWRITE; // Default rewrite method
+            }
+            return new PrefixQuery(new Term(name(), indexedValueForSearch(value)), method);
+        }
+
+        @Override
+        public Query fuzzyQuery(
+            Object value,
+            Fuzziness fuzziness,
+            int prefixLength,
+            int maxExpansions,
+            boolean transpositions,
+            MultiTermQuery.RewriteMethod method,
+            QueryShardContext context
+        ) {
+
+            if (method == null) {
+                method = MultiTermQuery.CONSTANT_SCORE_REWRITE;
+            }
+
+            return new FuzzyQuery(
+                new Term(name(), indexedValueForSearch(value)),
+                fuzziness.asDistance(),
+                prefixLength,
+                maxExpansions,
+                transpositions,
+                method
+            );
+        }
```
Let's wrap all these in `IndexOrDocValuesQuery`

harshavamsi commented on server/src/main/java/org/opensearch/index/mapper/SemanticVersionFieldMapper.java:
```diff
@@ -0,0 +1,313 @@
+        BytesRef normalizedValueBytes = new BytesRef(normalizedValue);
+
+        // For retrieval: store original version string
+        context.doc().add(new StoredField(fieldType().name(), versionString));
+
+        // For searching (term queries): use original version string
+        context.doc().add(new KeywordField(fieldType().name(), bytes, Field.Store.YES));
+
+        // For range queries and sorting: use normalized form
+        context.doc().add(new KeywordField(fieldType().name() + NORMALIZED_FIELD_SUFFIX, normalizedValueBytes, Field.Store.NO));
+        context.doc().add(new SortedSetDocValuesField(fieldType().name() + NORMALIZED_FIELD_SUFFIX, normalizedValueBytes));
```
Once we add the `hasDocValues` we would need to check if they are enabled before we store them

harshavamsi reviewed (commented):
Parsing of the Semantic Version looks alright to me but the FieldMapper needs some tweaks

deshsidd commented on server/src/main/java/org/opensearch/index/mapper/SemanticVersion.java:
```diff
@@ -0,0 +1,255 @@
+            version = version.replaceAll("\\s+", ".");
+        }
+
+        Pattern pattern = Pattern.compile(
```
Thanks, introduced a constant and a comment with the source.

deshsidd commented:
@harshavamsi @cwperks Thanks for the reviews. Made changes to refactor SemanticVersionFieldMapper for flexible index, store, and doc_values support and have addressed all the review comments.

deshsidd commented on test/framework/src/main/java/org/opensearch/search/aggregations/AggregatorTestCase.java:
```diff
@@ -213,6 +214,7 @@ public abstract class AggregatorTestCase extends OpenSearchTestCase {
         denylist.add(FieldAliasMapper.CONTENT_TYPE); // TODO support alias
         denylist.add(DerivedFieldMapper.CONTENT_TYPE); // TODO support derived fields
         denylist.add(StarTreeMapper.CONTENT_TYPE); // TODO evaluate support for star tree fields
+        denylist.add(SemanticVersionFieldMapper.CONTENT_TYPE); // TODO evaludate support for semantic version fields
```
Tests failing when I remove this. I have tested all the supported aggregators as part of the test I have added in the PR.
